### PR TITLE
[ESQL] Support date_nanos on functions that take "any" type

### LINKED ISSUE #109998: [ESQL] Support date nanos on functions that take any type
### Description

Most of these should be easy:

 - [x] Coalesce
 - [x] Case
 - [x] Greatest
 - [x] Least
 - [x] MV_COUNT
 - [x] MV_DEDUP
 - [x] MV_FIRST
 - [x] MV_LAST
 - [x] MV_MAX
 - [x] MV_MIN
 - [x] MV_SORT
 - [x] MV_SLICE

Most of the work in this ticket should just be wiring up tests, but that's still important to get right.

### PR DESCRIPTION
**Pull request by not-napoleon**
Resolves https://github.com/elastic/elasticsearch/issues/109998

For the most part, this is just adding tests.  `Greater` and `Least` have actual production code changes - notably `toEvaluator` is modified to map date nanos to the long evaluator.  This parallels the work done in https://github.com/elastic/elasticsearch/pull/113961.  I've added CSV tests and unit tests for all the functions listed in the original ticket.

### CODE REVIEW

**Comment by not-napoleon** [comment]
I've intentionally left the meta tests failing because I don't want to fix them here if we're about to drop them in https://github.com/elastic/elasticsearch/pull/113967.  If we decide not to go ahead with that, I can update the expected function definition checks in this PR.

**Review by ivancea — APPROVED** [review]
LGTM

**Review by astefan — APPROVED** [review]
LGTM

Worth adding `test-release` label to the PR or you are confident enough that it's ok without this check?

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/Greatest.java [hunk 1]
```diff
@@ -152,7 +152,7 @@ public ExpressionEvaluator.Factory toEvaluator(ToEvaluator toEvaluator) {
         if (dataType == DataType.INTEGER) {
             return new GreatestIntEvaluator.Factory(source(), factories);
         }
-        if (dataType == DataType.LONG || dataType == DataType.DATETIME) {
+        if (dataType == DataType.LONG || dataType == DataType.DATETIME || dataType == DataType.DATE_NANOS) {
```
Comment (bpintea) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/Greatest.java:155]: I guess that having a DataType method to check on the two date types would be useful (a la `isDateTimeOrNanos()`).

**Review by bpintea — APPROVED** [review]
Lgtm

**Comment by not-napoleon** [comment]
> Worth adding test-release label to the PR or you are confident enough that it's ok without this check?

Sure, I can add it.  Is that something we're just doing on every PR that is being developed behind a feature flag now?

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/Greatest.java (hunk 1 above)
Comment (not-napoleon) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/Greatest.java:155]: I really don't like adding more of those "helper" functions.  In my opinion, it's not any more readable than just having another case in the switching logic, and in fact may be less readable.  And I'd like to refactor this to use a map anyway, which we could reuse in the type checker.

**Comment by not-napoleon** [comment]
@elasticmachine update branch

**Comment by not-napoleon** [comment]
@elasticmachine update branch

**Comment by not-napoleon** [comment]
@elasticmachine update branch

**Comment by not-napoleon** [comment]
@elasticmachine update branch

**Pull request merged**